### PR TITLE
Make most imports explicit

### LINKED ISSUE #4: Use explicit imports

### PR DESCRIPTION
**Pull request by stu204146-kiel**
Most exports are explicit now

Closes #4

### CODE REVIEW

File: src/exe/Main.hs [hunk 1]
```diff
@@ -24,8 +39,8 @@ data Options = Options
     -- ^ The name of the output file or @Nothing@ if output should be printed
     --   to the console.
   , enableDebug  :: Bool
-    -- ^ Flag that indicates whether to print debugging messages to the console.
   , trivialCase  :: Bool
+  -- ^ Flag that indicates whether to print debugging messages to the console.
     -- ^ Flag that indicates whether to enable trivial case completion or not.
```
Comment (just95) [inline src/exe/Main.hs]: This comment somehow got moved.

File: src/lib/GuardElimination.hs [hunk 2]
```diff
@@ -9,45 +9,50 @@ where                                   -- TODO Apply GE to GuardedRhs in case e
                                                                                     -- not false by semantics
 
 import qualified Algo                          as A
+                                                ( err
+                                                , translatePVar
+                                                , newVars
+                                                , newVar
+                                                )
 import           Control.Monad
```
Comment (just95) [inline src/lib/GuardElimination.hs]: This import is missing the explicit import list.

File: src/test/AlgoTests.hs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 module AlgoTests where
 
 import           Control.Monad                  ( void )
-import           Language.Haskell.Exts
+import qualified Language.Haskell.Exts         as HSE
 import           Test.Hspec
```
Comment (just95) [inline src/test/AlgoTests.hs]: This import is also missing explicit imports.

**Review by just95 — CHANGES_REQUESTED** [review]

File: src/lib/Algo.hs [hunk 4]
```diff
@@ -195,37 +208,39 @@ findDataType cname = foldr
...
 
 -- | Creates new @case@ expression alternatives for the given missing
 --   constructors.
 createAltsFromConstr
-  :: Pat ()        -- ^ The fresh variable matched by the @case@ expression.
+  :: HSE.Pat ()        -- ^ The fresh variable matched by the @case@ expression.
   -> [Constructor] -- ^ The missing constructors to generate alternatives for.
-  -> Exp ()        -- ^ The error expression for pattern-matching failures.
-  -> PM [Alt ()]
+  -> HSE.Exp ()        -- ^ The error expression for pattern-matching failures.
```
Comment (stu204767) [inline src/lib/Algo.hs]: These comments should be aligned.

File: src/lib/Algo.hs [hunk 5]
```diff
@@ -40,10 +53,10 @@ err = Var () (UnQual () (Ident () "undefined"))
 --   All equations must have the same number of patterns as the given list
 --   of fresh variable patterns.
 match
-  :: [Pat ()] -- ^ Fresh variable patterns.
+  :: [HSE.Pat ()] -- ^ Fresh variable patterns.
   -> [Eqs]    -- ^ The equations of the function declaration.
-  -> Exp ()   -- ^ The error expression for pattern-matching failures.
-  -> PM (Exp ())
+  -> HSE.Exp ()   -- ^ The error expression for pattern-matching failures.
```
Comment (stu204767) [inline src/lib/Algo.hs]: These too.

File: src/lib/Algo.hs [hunk 6]
```diff
@@ -103,35 +116,35 @@ groupPat = groupBy ordPats
...
+isPVar _              = False
 
 -- | Applies 'match' to every group of equations where the error expression
 --   is the 'match' result of the next group.
 createRekMatch
-  :: [Pat ()] -- ^ Fresh variable patterns.
-  -> Exp ()   -- ^ The error expression for pattern-matching failures.
+  :: [HSE.Pat ()] -- ^ Fresh variable patterns.
+  -> HSE.Exp ()   -- ^ The error expression for pattern-matching failures.
   -> [[Eqs]]  -- ^ Groups of equations (see 'groupPat').
```
Comment (stu204767) [inline src/lib/Algo.hs]: This one too.
```suggestion
  -> [[Eqs]]      -- ^ Groups of equations (see 'groupPat').
```

File: src/lib/Algo.hs [hunk 7]
```diff
@@ -103,35 +116,35 @@ groupPat = groupBy ordPats
...
   foldr (\eqs mrhs -> mrhs >>= match vars eqs) (return er)
 
 -- | Creates a case expression that performs pattern matching on the variable
 --   bound by the given variable pattern.
 makeRhs
-  :: Pat ()   -- ^ The fresh variable pattern to match.
-  -> [Pat ()] -- ^ The remaing fresh variable patterns.
+  :: HSE.Pat ()   -- ^ The fresh variable pattern to match.
+  -> [HSE.Pat ()] -- ^ The remaing fresh variable patterns.
   -> [Eqs]    -- ^ The equations.
```
Comment (stu204767) [inline src/lib/Algo.hs]: This one as well.
```suggestion
  -> [Eqs]        -- ^ The equations.
```

File: src/lib/Algo.hs [hunk 8]
```diff
@@ -144,11 +157,11 @@ translatePVar p =
 --   If trivial case completion is not enabled, one alternative is added for
 --   every missing constructor.
 computeAlts
-  :: Pat ()   -- ^ The variable pattern that binds the matched variable.
-  -> [Pat ()] -- ^ The remaing fresh variable patterns.
+  :: HSE.Pat ()   -- ^ The variable pattern that binds the matched variable.
+  -> [HSE.Pat ()] -- ^ The remaing fresh variable patterns.
   -> [Eqs]    -- ^ The equations to generate alternatives for.
```
Comment (stu204767) [inline src/lib/Algo.hs]:
```suggestion
  -> [Eqs]        -- ^ The equations to generate alternatives for.
```

File: src/lib/Algo.hs [hunk 9]
```diff
@@ -316,11 +332,11 @@ consName pat =
 --   termination check of the free-compiler and can probably be removed
 --   once sharing is implemented.
 computeAlt
-  :: Pat ()   -- ^ The variable pattern that binds the matched variable.
-  -> [Pat ()] -- ^ The remaing fresh variable patterns.
-  -> Exp ()   -- ^ The error expression for pattern-matching failures.
+  :: HSE.Pat ()   -- ^ The variable pattern that binds the matched variable.
+  -> [HSE.Pat ()] -- ^ The remaing fresh variable patterns.
+  -> HSE.Exp ()   -- ^ The error expression for pattern-matching failures.
   -> [Eqs]    -- ^ A group of equations (see 'groupByCons').
```
Comment (stu204767) [inline src/lib/Algo.hs]:
```suggestion
  -> [Eqs]        -- ^ A group of equations (see 'groupByCons').
```

File: src/lib/GuardElimination.hs [hunk 10]
```diff
@@ -9,45 +9,50 @@ where                                   -- TODO Apply GE to GuardedRhs in case e
...
 -- Folds the list of GExps to declarations.
 foldGEqs
-  :: [Pat ()]               -- fresh variables for the case exps
-  -> ([Decl ()], Pat ())    -- startcase ([], first generated Pattern)
+  :: [HSE.Pat ()]               -- fresh variables for the case exps
+  -> ([HSE.Decl ()], HSE.Pat ())    -- startcase ([], first generated Pattern)
   -> [GExp]                 -- list of pattern + rhs pair
-  -> PM ([Decl ()], Pat ()) -- a list of declarations for the let binding and a
+  -> PM ([HSE.Decl ()], HSE.Pat ()) -- a list of declarations for the let binding and a
                             -- free Variable for the error case
```
Comment (stu204767) [inline src/lib/GuardElimination.hs]: These comments should be aligned as well.

File: src/lib/Algo.hs [hunk 11]
```diff
@@ -18,20 +18,33 @@ where
 
 import           Data.List
 import           Data.Function
```
Comment (stu204767) [inline src/lib/Algo.hs]: Shouldn't these be explicit as well?
```suggestion
import           Data.List                      ( partition
                                                , groupBy
                                                )
import           Data.Function                  ( on )
```

File: src/lib/GuardElimination.hs [hunk 12]
```diff
@@ -9,45 +9,50 @@ where                                   -- TODO Apply GE to GuardedRhs in case e
...
 
 -- Generates a varbinding and a new variable for the next var binding
 createDecl
-  :: [Pat ()]            -- generated varibles
-  -> ([Decl ()], Pat ()) -- (current decls , variable for let binding)
+  :: [HSE.Pat ()]            -- generated varibles
+  -> ([HSE.Decl ()], HSE.Pat ()) -- (current decls , variable for let binding)
   -> GExp                -- pairs of pattern to match against and a guarded Rhs
-  -> PM ([Decl ()], Pat ()) -- var bindings , variable for next match
+  -> PM ([HSE.Decl ()], HSE.Pat ()) -- var bindings , variable for next match
```
Comment (stu204767) [inline src/lib/GuardElimination.hs]: These comments should be aligned.

**Review by stu204767 — CHANGES_REQUESTED** [review]

**Review by just95 — APPROVED** [review]

**Review by stu204767 — APPROVED** [review]

**Pull request merged**